Sources/AuthFoundation/Utilities/CoalescedResult.swift

```diff
@@ -12,22 +12,106 @@
+///
+/// This is commonly used to limit unnecessary duplication of network operations for common resources.
+@_documentation(visibility: private)
+public actor CoalescedResult<T: Sendable>: Sendable {
+    private let taskName: String?
+    private var task: BackgroundTask?
+    private var continuations: [CheckedContinuation<T, Error>] = []
+    private var _value: T?
+    private var _isActive: Bool = false  {
+        didSet {
+            if _isActive, let taskName {
+                task = BackgroundTask(named: taskName)
+            } else if let task {
+                task.finish()
+                self.task = nil
+            }
+        }
+    }
+
+    /// Designated initializer.
+    /// - Parameter taskName: The name of a background task to initiate while the operation is active.
+    public init(taskName: String? = nil) {
+        self.taskName = taskName
     }
-    
-    func start(_ operation: ((T) -> Void) -> Void) {
-        operation { result in
-            self.finish(result)
+
+    /// Indicates if the asynchronous operation is being performed.
+    nonisolated public var isActive: Bool {
+        let semaphore = DispatchSemaphore(value: 0)
+        nonisolated(unsafe) var result: Bool = false
+        Task {
+            result = await self._isActive
+            semaphore.signal()
         }
+        semaphore.wait()
+        return result
     }
 
-    func finish(_ result: T) {
-        completionHandlers.forEach { completion in
-            completion(result)
+    /// Stores the previously-fetched value, if one has been fetched and its response was successful.
+    nonisolated public var value: T? {
+        let semaphore = DispatchSemaphore(value: 0)
+        nonisolated(unsafe) var result: T?
+        Task {
+            result = await self._value
+            semaphore.signal()
+        }
+        semaphore.wait()
+        return result
+    }
+
+    /// Performs an asynchronous operation, while ensuring that multiple concurrent requests do not result in multiple active tasks.
+    /// - Parameters:
+    ///   - reset: Resets any previously fetched value before starting.
+    ///   - operation: Asynchronous operation used to fetch the value.
+    ///   - willBegin: Optional closure to invoke before beginning the operation.
+    ///   - didEnd: Optional closure to invoke with the result after the operation completes.
+    /// - Returns: The value of type `T`.
+    public func perform(reset: Bool = false,
+                        operation: () async throws -> T,
+                        willBegin: () -> Void = {},
+                        didEnd: (Result<T, Error>) -> Void = { _ in }) async throws -> T
+    {
+        if reset {
+            _value = nil
+        }
+
+        if !_isActive {
+            if let value = _value {
+                return value
+            }
+
+            _isActive = true
+            task = BackgroundTask(named: taskName)
+            willBegin()
```

Comment:
Is it fine that begin and end will be called only for the first caller?

Reply:
Yes, these should only be called once, regardless of the number of callers invoking it.  This is typically because `willBegin` and `didEnd` trigger calls to delegates, and duplicated notifications can cause problems for delegate consumers.